radix-transactions/src/manifest/static_resource_movements/visitor.rs

```diff
@@ -0,0 +1,724 @@
+                        fungible_resource_address,
+                        FungibleBounds {
+                            lower: LowerFungibleBound::Amount(bucket_amount),
+                            upper: UpperFungibleBound::Amount(bucket_amount),
+                        },
+                    ),
+                );
+            }
+            // Taking an amount from the worktop of a non-fungible. We can't tell which non-fungible
+            // ids are being taken so the bucket will just contain unknown ids.
+            (
+                CompositeResourceAddress::NonFungible(non_fungible_resource_address),
+                BucketSourceAmount::AmountFromWorktop {
+                    amount: bucket_amount,
+                    ..
+                },
+            ) => {
+                // Worktop accounting.
+                if let Some(worktop_non_fungible_content) = self
+                    .worktop_non_fungible_contents
+                    .get_mut(&non_fungible_resource_address)
+                {
+                    // Reduce the amount bounds.
+                    worktop_non_fungible_content
+                        .amount_bounds
+                        .decrease_both_bounds(bucket_amount);
+                    // Switch the id bounds to be unknown.
+                    worktop_non_fungible_content.id_bounds = NonFungibleIdBounds::Unknown;
+                }
+                // Creation of bucket.
+                self.tracked_buckets.insert(
+                    bucket,
+                    BucketContent::NonFungible(
+                        non_fungible_resource_address,
+                        NonFungibleBounds {
+                            amount_bounds: FungibleBounds::new_exact(bucket_amount),
+                            id_bounds: NonFungibleIdBounds::Unknown,
+                        },
+                    ),
+                );
+            }
+            // Taking non-fungibles from the worktop by id.
+            (
+                CompositeResourceAddress::NonFungible(non_fungible_resource_address),
+                BucketSourceAmount::NonFungiblesFromWorktop {
+                    ids: bucket_ids, ..
+                },
+            ) => {
+                let bucket_ids = bucket_ids.iter().cloned().collect::<IndexSet<_>>();
+                let bucket_ids_amount = Decimal::from(bucket_ids.len());
+
+                // Worktop accounting.
+                if let Some(worktop_non_fungible_content) = self
+                    .worktop_non_fungible_contents
+                    .get_mut(&non_fungible_resource_address)
+                {
+                    // Reduce the amount bounds.
+                    worktop_non_fungible_content
+                        .amount_bounds
+                        .decrease_both_bounds(bucket_ids_amount);
+
+                    // Remove the ids from the set of ids in the worktop.
+                    if let NonFungibleIdBounds::FullyKnown(ref mut id_bounds)
```

Comment:
(Addressed in https://github.com/radixdlt/radixdlt-scrypto/pull/1931 )

Reply:
This depends on whether the worktop has an uncertainty source or not.

Reply:
> This depends on whether the worktop has an uncertainty source or not.

Two things:
* In my example, there is no uncertainty source, and I'm saying in this case logic is incorrect.
* What does an uncertainty source mean in your model? Does it mean that any resource doesn't have an upper bound? Or just unspecified sources?

My model correctly allows this (because it's possible):
* Withdraw A, B, C from NF X
* Withdraw 1 NF X
* Take from worktop B, D
* ...

And fails this (line 3 is impossible, because it would leave A and C in the bucket with a max balance of 1):
* Withdraw A, B, C from NF X
* Withdraw 1 NF X
* Take from worktop B, D, E
* ...